solutions/platinum/izho-19-HedgehogDaniyarAndAlgorithms.mdx

```diff
@@ -0,0 +1,184 @@
+---
+id: izho-19-HedgehogDaniyarAndAlgorithms
+source: IZhO
+title: 2019 - Hedgehog Daniyar and Algorithms
+author: Justin Ji
+---
+
+<Spoiler title = "Hint">
+
+The number of inversions in an array is the number of adjacent
+swaps we need to make to sort an array in non-decreasing order.
+How can we use this idea to find the minimum value of $k$ required to sort
+the array?
+
+</Spoiler>
+
+<Spoiler title = "Explanation">
+
+In our array, we only need to swap elements that are inversions. That is
+to say, if we have a pair $(i, j)$ where $i < j$ and $w_i > w_j$, then we
+must swap them when sorting the array. Thus, the answer is the maximum
+value of $w_i + w_j$ among all inversion pairs.
+
+To answer the queries, we sweep right to left on the left endpoint whilst
+maintaining some data structure to keep track of the minimum value of $k$ for
+all the right endpoints. One key observation to make is that if at our current
+index $i$ there exists a $j$ where $i < j$ and $w_i \leq w_j$, then we can't
+increase the value of any inversion pairs where the second value's index is
+greater than $j$. Thus, we only need to update the inversion pair
+values on the range $[i + 1, j - 1]$. To find the said value of $j$, we utilize a
+monotonic stack.
+
+The most straightforward way to solve the problem from here is to use a lazy
+segment tree to keep track of the best inversion pair value for each index
+(where the value of an inversion pair $(i, j)$ is stored at index $j$)
+and the best inversion pair value on a range. To implement this, notice how
+every inversion pair value stored at an index $j$ is $w_j + x$, where $x$ is
+some value $w_i$. Thus, we can use a range update to apply the best value of $x$.
+
+Note that a more elegant (although less intuitive) Binary Indexed Tree solution
+also exists. For that solution, we utilize a BIT to keep track of the prefix
+maximum value and update it while popping items in our monotonic stack.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}((N + Q)\log{N})$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+template <class Info, class Tag> class LazySegtree {
```

Comment:
ok, _now_ is this consistent w the rurq one
if so you can codesnip this

Reply:
done